### DISCUSSION TITLE
Documentation for extensions

### DISCUSSION THREAD

**Discussion by simonh1000** [remarkjs/remark#837] [Q&A] [answered]
Is there any official documentation for extensions? 

I've followed the example, and add some notes of what I could grasp, but understanding the full API (enter, exit, ...) would be very useful. The code below matches `{xyz}` but I'm unclear how - for example to convert that to `{{xyz}}`

```
// is name here deliberately the same as the first effects.enter?
const remixConstruct = {name: 'remix', tokenize: remixTokenize}

// extension name is 'remix'
// text construct for a '{' (123)
export const remix = {text: {123: remixConstruct}}

function remixTokenize(effects, ok, nok) {
    return start

    function start(code) {
        effects.enter('remix')
        effects.enter('remixMarker')
        effects.consume(code)
        effects.exit('remixMarker')
        effects.enter('remixString')
        // next line essential, but not sure why
        effects.enter('chunkString', {contentType: 'string'})
        return begin
    }

    // if the first character we find is '}', then we must abandon
    function begin(code) {
        return code === 125 ? nok(code) : inside(code)
    }

    // we know the first character is NOT '}'
    // we continue consuming until we get to '}'
    function inside(code) {
        // 'abandon' if we find these ones
        if (code === -5 || code === -4 || code === -3 || code === null) {
            return nok(code)
        }

        // found the '}'
        if (code === 125) {
            effects.exit('chunkString')
            effects.exit('remixString')
            effects.enter('remixMarker')
            effects.consume(code)
            effects.exit('remixMarker')
            effects.exit('remix')
            return ok
        }

        effects.consume(code)
        return inside
    }
}

// mdastExtensions - is this the compile stage?
// do something with the content collected by effects.enter('remixString') ... effects.exit('remixString')
// see how the keys to enter, and exit
export function remixMdast() {
    return {
        enter: {'remixString': enterRemixString},
        exit: {'remixString': exitRemixString},
    }

    function enterRemixString(token) {
        this.enter({type: 'inline', children: []}, token)
    }

    function exitRemixString(token) {
        this.exit(token)
    }
}
```

**Comment by ChristianMurphy** [comment, marked as answer]
> Documentation for extensions
> understanding the full API (enter, exit, ...) would be very useful

You appear to want to add a syntax extension?
See https://github.com/micromark/micromark#extensions

> I've followed the example

Which example are you following?

> The code below matches `{xyz}`

It sounds like you are going for JSX support?
Consider checking out https://github.com/wooorm/xdm

**Reply by simonh1000** [reply]
Not JSX, this will be totally proprietary.  I need to be able to identify parts of a text, for example

`the quick {{brown fox}} jumped over the ...`

should result in 3 children, split before and after the text in `{{ }}`

**Reply by ChristianMurphy** [reply]
:man_shrugging: it's quite similar to several existing plugins.
Including: https://github.com/djm/remark-shortcodes, https://github.com/josemarluedke/remark-hbs, and https://github.com/remarkjs/remark-directive

As https://github.com/micromark/micromark#extensions notes, syntax extensions are supported and are documented. But generally using an existing standard is a better idea than creating a new one.
